### ISSUE TITLE
incorrect results in unionfind compared to other implementations.

### ISSUE THREAD

**Issue by tommyjcarpenter** [networkx/networkx#3567]
I am seeing an issue where I need to reinsert all of my original keys, during an algorithm that continuously calls `union`. The identical code, using other implementations of UnionFind, work, but using the networkx union find leads to incorrect results. 

This is a solution to a coursera problem, using the input file at: https://d18ky98rnyall9.cloudfront.net/_fe8d0202cd20a808db6a4d5d06be62f4_clustering1.txt?Expires=1567641600&Signature=OBpjnbzYQtcwHrFL97bXYvUeS2RD4BVkpNbmPCqJcwF2KPoY2WM5KLg5YofiG8OJ-A64qT8Jo02-6YkH7r5xcaXJghbCgbxecB31m3PzjvMeO5pzsiUbtUmsWLcFXeJqbpYsmraPucoXbM~PpWKi6CLdHEaQVRfB-LeYlDHWX-s_&Key-Pair-Id=APKAJLTNE6QMUY6HBC5A

The solution shrinks a network until there are only 4 clusters, then it finds the smallest edge between the clusters.

Here is my working solution runnable using that file (only the code in fix and cluster is relevant here). The problem is the calls to `fix` which should only be needed outside of the loop (once to initialize, since the package doesn't have an explicit insert), and should not be needed inside the loop. I have verified this using another UF implementation using the exact same code (See comment below).

My guess of what's happening is that `union` is butchering the keyspace somehow, but reinserting the original keys that were unioned fixes the bug for some reason. 

```
from networkx.utils import UnionFind


def find_minimal(uf, edges):
    # now we have to search for the lowest cost edge that connects them
    m = 9999999999999
    for i in range(1, 501):
        for j in range(1, 501):
            if i != j and uf[i] != uf[j]:
                for eindex in start_at[i]:
                    if edges[eindex][1] == i and edges[eindex][2] == j and edges[eindex][3] < m:
                        m = edges[eindex][3]
                        print(m)


def fix(uf):
    for k in range(1, 501):
        uf[k]


def cluster(edges, start_at):
    uf = UnionFind()

    sortede = sorted(edges, key=lambda x: x[3])

    fix(uf)

    for _, u, v, w in sortede:

        # moving this outside of the block makes this not work?????????
        fix(uf) #!!!!!!!

        if len(list(uf.to_sets())) == 4:
            break
        elif uf[u] != uf[v]:
            uf.union(u, v)

    find_minimal(uf, edges)


if __name__ == "__main__":
    edges = []
    start_at = {}
    with open("course_3_p2q1.txt", "r") as f:
        for lindex, line in enumerate(f):
            i = line.split()
            start = int(i[0])
            end = int(i[1])
            edges.append((lindex, start, end, int(i[2])))
            if start not in start_at:
                start_at[start] = []
            if end not in start_at:
                start_at[end] = []
            # we need these later when we need to quickly scan through edjes that connect clusters
            start_at[start].append(lindex)
            start_at[end].append(lindex)

    cluster(edges, start_at)
```

If I don't reinsert all keys on every iteration; that is, if I remove the `fix(uf)` INSIDE the for loop, I get the complete wrong answer, and at the end, uf.parents is different than the above working solution.

My guess is that the merging algorithm is blowing away some kind of historical membership.

Any ideas?

**Comment by tommyjcarpenter** [comment]
This is definitely a bug in networkx; switching the algorithm to https://raw.githubusercontent.com/deehzee/unionfind/master/unionfind.py

verbatim, without the fix, works.

```
from unionfind import UnionFind


def find_minimal(uf, edges):
    # now we have to search for the lowest cost edge that connects them
    m = 9999999999999
    for i in range(1, 501):
        for j in range(1, 501):
            if i != j and not uf.connected(i, j):
                for eindex in start_at[i]:
                    if edges[eindex][1] == i and edges[eindex][2] == j and edges[eindex][3] < m:
                        m = edges[eindex][3]
                        print(m)


def cluster(edges, start_at):
    uf = UnionFind()

    sortede = sorted(edges, key=lambda x: x[3])

    for k in range(1, 501):
        uf.add(k)

    for _, u, v, w in sortede:

        if len(list(uf.components())) == 4:
            break
        elif not uf.connected(u, v):
            uf.union(u, v)

    find_minimal(uf, edges)


if __name__ == "__main__":
    edges = []
    start_at = {}
    with open("course_3_p2q1.txt", "r") as f:
        for lindex, line in enumerate(f):
            i = line.split()
            start = int(i[0])
            end = int(i[1])
            edges.append((lindex, start, end, int(i[2])))
            if start not in start_at:
                start_at[start] = []
            if end not in start_at:
                start_at[end] = []
            # we need these later when we need to quickly scan through edjes that connect clusters
            start_at[start].append(lindex)
            start_at[end].append(lindex)

    cluster(edges, start_at)
```

**Comment by dschult** [comment]
I cannot reproduce this result. The link to the edgelist file is no longer active. I have tried to recreate it with randomly generated edgelists. I have also tried looking through the logic of the algorithm to see what might be going wrong. 

The first thing done in ```uf.to_sets()``` is essentially what is in this Issue's ```fix(uf)``` function, except it is done with all nodes instead of nodes 1 to 500.  That means there should be no difference between the code with ```fix(uf)``` inside the loop and outside the loop. 

I suspect something else is/was going wrong to make it appear to fix the problem.  
Can anyone reproduce any kind of error with this info?

**Comment by tommyjcarpenter** [comment]
Solution: https://github.com/tommyjcarpenter/coursera-algorithms/blob/master/src/courseraalg/course_3_p2q1.py

Input file mentioned above: https://github.com/tommyjcarpenter/coursera-algorithms/blob/master/src/courseraalg/course_3_p2q1.txt

swapping out the "unionfind" package with this package leads to the incorrect results, given the rest of the code verbatim

**Comment by tommyjcarpenter** [comment]
you should be able to download those two files and run them directly, then try swapping out the networkx package in the first file with yours and running again

**Comment by boothby** [comment]
I haven't dug into the example, but I read over `unionfind` and found a bug in union:

```
    def union(self, *objects):
        """Find the sets containing the objects and merge them all."""
        roots = [self[x] for x in objects]
        # Find the heaviest root according to its weight.
        heaviest = max(roots, key=lambda r: self.weights[r])
        for r in roots:
            if r != heaviest:
                self.weights[heaviest] += self.weights[r]
                self.parents[r] = heaviest
```

If `roots` has a repeated element (other than `r`, then that root's `weight` will get accumulated too much.  This can be fixed by turning the list-comprehension into a set.

**Comment by dschult** [comment]
I have downloaded both versions of the solution and both the couresra unionfind.py and the networkx union_find.py give the same output.

I notice that in NetworkX v2.4 there is a fix of union_find.py for set extraction. #3224 
@tommyjcarpenter can you verify that you were using NX v2.3 or before?

**Comment by tommyjcarpenter** [comment]
I was using whatever the latest version in pip was at the time of this issue. So, sep 3 2019. It is perhaps possible that it wasn't in pypi at that time and maybe I had an older version? I don't have the repo or the venv anymore to check..

**Comment by dschult** [comment]
OK... That would have been v2.3.
I have now tested that it is broken on v2.3 and fixed on v2.4 and the github dev branch.

@tommyjcarpenter   Thanks for posting this so nicely and completely!!

It also helped give @boothby an incentive to look into unionfind and discover a bug in ```union```.
The bug in union, if I understand correctly, leads to an inefficiency in the tree structure.
I believe it still gave correct results, just didn't choose the best tree structure for future lookups.

If this is the case, then we can just replace the list comprehension with a set comprehension.

**Comment by tommyjcarpenter** [comment]
It looks like your fix got merged in May, but if there was a delay pushing to pypi, or a release came after that, that makes sense!

Glad we got this figured out.  I actually spent several hours on that assignment due to this issue lol, I kept re-reading the code over and over wondering what the issue was.. finally it dawned on me that maybe the union find structure wasn't working correctly.

Thanks for your hard work - Tommy
